docs/design/05-asset-layout.md

```diff
@@ -0,0 +1,147 @@
+directory, `/v/l/m/asset-layouts`. Also the asset layouts will not be
+loaded into the namespace.
+
+## Editing of a layout
+
+Editing of an existing asset `layout` will be the same as an existing
+`asset`. To implement this, a [`-l`, `--layout`] flag will be added to
+the `metal asset edit` command. Otherwise it is exactly the same.
+
+## Creating an Asset from a Layout
+
+The primary purpose of the `layouts` is the ability to rapidly generate
+assets from them. The [`-l`, `--layout`] flag needs to be added to the
+`metal asset add` command. This will tell the command to add a new
+asset from the `layout`. As each `layout` has a type, this will be the
+type of the asset. This may not be required if `types` and `layouts`
+have unique names.
+
+### Generation of Sub Assets
+
+The `^name` notation in the `layouts` is going to be slightly different
+then `assets`. In the assets, `^name` denotes a link to another 
+asset. In the layouts, the notation going to be:
+`type_or_layout_name^base_asset_name`.
+
+Once the asset (from a layout) has been edited, the `yaml` is parsed
+for the above notation. This notation tells metalware to create an 
+asset of type: `type_or_layout_name`. Asset layouts and types have
+unique names, it is possible to tell which one should be used.
+
+Then the `base_asset_name` tells metalware how to construct the
+translated asset name. As layouts are going to be used multiple 
+times, they can not all use the same name. Thus the syntax for the
+sub asset name will be: `#{parent_asset_name}_#{base_asset_name}`.
+
+The penultimate step of creating an asset form the layout is to
+replace `type_or_layout_name^base_asset_name` with
+`^#{parent_asset_name}_#{base_asset_name}` and thus automatically
+creates the link between the asset and sub assets.
+
+The final step is then to prompt the user if they want to create each
+sub asset, if so the process is repeated.
+
+### Key questions:
+
+1. Should the sub-assets be made in depth first or breadth first order
+2. When a sub asset is created, should the `^parent_asset` link be
+   automatically generated?
+   * This could be done using the parents asset type as the key
+3. How are sub-assets handled if they already exists? Should it just
+   be an edit? How does point 2 affect this?
+4. In a layout, should there be special handling of `^asset_name`
+   notation? There has been discussion of using the hash key that
+   contains it as the `type` for the auto generation of sub assets.
+   This, however, removes the ability to statically reference an
+   asset from a `layout`.
+
+
+## Side Issue - Temp Filename
+
+When the editor is open, it is hard to know which asset is currently
+be edited because of the use of a temporary file.
+
+The temp file should still be used, however its name should be updated
+to include the final save name NOT the asset type. Also the random bit
+of the name should be at the end. Thus the temp file path should be:
+`/tmp/asset_name.yaml.XXXXXXXX`
```

Comment:
Giving it a name like this is likely to mess up automatic syntax highlightling in various editors, as a `*.yaml.XXXXXXXX` file is not a `yaml` file. You could just include the `XXXX` part in a different part of the file name, or instead have a directory where all asset editing is done. Something like `/var/lib/metalware/tmp/assets/`, and then use the original asset names within this.

If assets were then `cp`ed rather than `mv`ed into place this would also have the advantage of leaving old invalid asset files in place where they can be checked.